---
title: "Downloading protein and DNA sequences with the rentrez package"
description: |
  `rentrez` is a fast and easy way to access biological sequence databases.
author: Nathan L. Brouwer
date: 2022-05-12
output:
  distill::distill_article:
    self_contained: false
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

shroom_table <- c("CAA78718" , "X. laevis Apx" ,         "xShroom1",
                  "NP_597713" , "H. sapiens APXL2" ,     "hShroom1",
                  "CAA58534" , "H. sapiens APXL",        "hShroom2",
                  "ABD19518" , "M. musculus Apxl" ,      "mShroom2",
                  "AAF13269" , "M. musculus ShroomL" ,   "mShroom3a",
                  "AAF13270" , "M. musculus ShroomS" ,   "mShroom3b",
                  "NP_065910", "H. sapiens Shroom" ,     "hShroom3",
                  "ABD59319" , "X. laevis Shroom-like",  "xShroom3",
                  "NP_065768", "H. sapiens KIAA1202" ,   "hShroom4a",
                  "AAK95579" , "H. sapiens SHAP-A" ,     "hShroom4b",
                 #"DQ435686" , "M. musculus KIAA1202" ,  "mShroom4",
                  "ABA81834" , "D. melanogaster Shroom", "dmShroom",
                  "EAA12598" , "A. gambiae Shroom",      "agShroom",
                  "XP_392427" , "A. mellifera Shroom" ,  "amShroom",
                  "XP_783573" , "S. purpuratus Shroom" , "spShroom") #sea urchin
# What we just made is just one long vector with all the info.
# is(shroom_table)
# class(shroom_table)
# length(shroom_table)


# I'll do a bit of formatting; you can ignore these details if you want
# convert the vector to matrix using matrix()
shroom_table_matrix <- matrix(shroom_table,
                                  byrow = T,
                                  nrow = 14)

# convert the matrix to a dataframe using data.frame()
shroom_table <- data.frame(shroom_table_matrix, 
                     stringsAsFactors = F)

# name columns of dataframe using names() function
names(shroom_table) <- c("accession", "name.orig","name.new")

# Create simplified species names
## access species column using $ notation
shroom_table$spp <- "Homo"
shroom_table$spp[grep("laevis",shroom_table$name.orig)] <- "Xenopus"
shroom_table$spp[grep("musculus",shroom_table$name.orig)] <- "Mus"
shroom_table$spp[grep("melanogaster",shroom_table$name.orig)] <- "Drosophila"
shroom_table$spp[grep("gambiae",shroom_table$name.orig)] <- "mosquito"
shroom_table$spp[grep("mellifera",shroom_table$name.orig)] <- "bee"
shroom_table$spp[grep("purpuratus",shroom_table$name.orig)] <- "sea urchin"


```


<!-- Phylogenies play an important role in computational biology and bioinformatics.  Phylogenetics itself is an obligatly computational field that only began rapid growth when computational power allowed the many algorithms it relies on to be done rapidly.  Phylogeneies of species, genes and proteins are used to address many biological issues, including -->

<!-- * Patterns of protein evolution -->
<!-- * Origin and evolution of phenotypic traits -->
<!-- * Origin and progression of epidemics -->
<!-- * Origin of evolution of diseases (e.g., zooenoses) -->
<!-- * Prediction of protein function from its sequence -->
<!-- * ... and many more -->

<!-- The actual building of a phylogeny is a computationally intensive task; moreover, there are many bioinformatics and computational tasks the precede the construction of a phylogeny: -->

<!-- * genome sequencing and assembly -->
<!-- * computational gene prediction and annotation -->
<!-- * database searching and results screening -->
<!-- * pairwise sequence alignment -->
<!-- * data organization and cleaning -->
<!-- * multiple sequence alignment -->
<!-- * evaluation and validation of alignment accuracy -->

<!-- Once all of these steps have been carried out, the building of a phylogeny involves -->

<!-- * picking a model of sequence evolution or other description of evolution -->
<!-- * picking a statistical approach to tree construction -->
<!-- * evaluating uncertainty in the final tree -->

<!-- In this chapter we will work through many of these steps.  In most cases we will pick the easiest or fastest option; in later chapters we will unpack the various options.  This chapter is written as an interactive R session.  You can follow along by opening the .Rmd file of the chapter or typing the appropriate commands into your own script.  I assume that all the necessary packages have been installed and they only need to be loaded into R using the *library()* command. -->


<!-- This lesson walks you through and entire workflow for a bioinformatics project, including -->

<!-- <!-- check all of this --> -->

<!-- 1. obtaining FASTA sequences -->
<!-- 1. cleaning sequences -->
<!-- 1. creating alignments -->
<!-- 1. creating a distance matrix -->
<!-- 1. building a phylogenetic tree -->

<!-- We'll examine the Shroom family of genes, which produces Shroom proteins essential for tissue formation in many multicellular eukaryotes, including neural tube formation in vertebrates.  We'll examine shroom in several very different organism, including humans, mice and sea urchins.  There is more than one type of shroom in vertebrates, and we'll also look at two different Shroom genes: shroom 1 and shroom 2. -->

<!-- This lesson draws on skills from previous sections of the book, but is written to act as a independent summary of these activities.  There is therefore review of key aspects of R and bioinformatics throughout it. -->


<!-- ### Vocab  -->

<!-- ## Software Preliminaries -->

<!-- * argument -->
<!-- * function -->
<!-- * list -->
<!-- * named list -->
<!-- * vector -->
<!-- * named vector -->
<!-- * for() loop -->
<!-- * R console -->



<!-- ### R functions -->

<!-- * library() -->
<!-- * round() -->
<!-- * plot() -->
<!-- * mtext() -->
<!-- * nchar() -->
<!-- * rentrez::entrez_fetch() -->
<!-- * compbio4all::entrez_fetch_list() -->
<!-- * compbio4all::print_msa() (Coghlan 2011) -->
<!-- * Biostrings::AAStringSet() -->
<!-- * msa::msa() -->
<!-- * msa::msaConvert() -->
<!-- * msa::msaPrettyPrint() -->
<!-- * seqinr::dist.alignment() -->
<!-- * ape::nj() -->

<!-- A few things need to be done to get started with our R session. -->

<!-- ### Download necessary packages -->

<!-- Many *R* sessions begin by downloading necessary software packages to augment *R's* functionality. -->

<!-- If you don't have them already, you'll need the following packages from CRAN: -->

<!-- 1. `ape` -->
<!-- 1. `seqinr` -->
<!-- 1. `rentrez` -->
<!-- 1. `devtools` -->

<!-- The CRAN packages can be loaded with `install.packages()`.   -->

<!-- You'll also need these packages from Bioconductor: -->

<!-- 1. `msa` -->
<!-- 1. `Biostrings` -->

<!-- For installing packages from Bioconductor, see the chapter at the beginning of this book on this process. -->

<!-- Finally, you'll need these package from GitHub -->

<!-- 1. `compbio4all` -->
<!-- 1. `ggmsa` -->


<!-- To install packages from GitHub you can use the code `devtools::install_github("brouwern/compbio4all")` and `devtools::install_github("YuLab-SMU/ggmsa")` -->


<!-- This code chunk downloads most of these packages.  The code has been commented out because it only needs to be run once.   -->

<!-- ```{r} -->
<!-- # Install needed packages -->

<!-- ## CRAN PACKAGES -->
<!-- # downloaded using install.packages() -->
<!-- # install.packages("rentrez",dependencies = TRUE) -->
<!-- # install.packages("devtools") -->
<!-- # install.packages("ape") -->
<!-- # install.packages("seqinr") -->

<!-- ### BiocManager - CRAN package to download  -->
<!-- ##### Bioconductor packages -->
<!-- # requires BiocManager -->
<!-- # install.packages("BiocManager") -->

<!-- ## BioConductor packages -->
<!-- ### downloaded with BiocManager::install(), NOT install.packages() -->
<!-- # BiocManager::install("msa") -->
<!-- # BiocManager::install("Biostrings") -->


<!-- ## GitHub packages -->
<!-- ### requires devtools package and its function install_github() -->
<!-- # library(devtools) -->
<!-- # devtools::install_github("brouwern/combio4all") -->
<!-- # devtools::install_github("YuLab-SMU/ggmsa") -->
<!-- ``` -->

<!-- ### Load packages into memory -->

<!-- We now need to load up all our bioinformatics and phylogenetics software into R.  This is done with the `library()` command.   -->

<!-- To run this code just click on the sideways green triangle all the way to the right of the code. -->

<!-- NOTE: You'll likely see some red code appear on your screen.  No worries, totally normal! -->


<!-- ```{r, message= F, warning=F} -->
<!-- # github packages -->
<!-- library(compbio4all) -->
<!-- library(ggmsa) -->

<!-- # CRAN packages -->
<!-- library(rentrez) -->
<!-- library(seqinr) -->
<!-- library(ape) -->

<!-- # Bioconductor packages -->
<!-- ## msa -->
<!-- ### The msa package is having problems on some platforms -->
<!-- ### You can skip the msa steps if necessary.  The msa output -->
<!-- ### is used to make a distance matrix and then phylogenetics trees, -->
<!-- ### but I provide code to build the matrix by hand so -->
<!-- ### you can proceed even if msa doesn't work for you. -->
<!-- library(msa) -->

<!-- ## Biostrings -->
<!-- library(Biostrings) -->
<!-- ``` -->




<!-- start blogplost 1 -->



## Downloading macro-molecular sequences

```{r, echo = F}
im <- magick::image_read("US-NLM-NCBI-Logo.png")
im <- magick::image_scale(im, "200")
magick::image_write(im, "US-NLM-NCBI-Logo2.png")
knitr::include_graphics("US-NLM-NCBI-Logo2.png")
```

In this post we'll do 2 things:

1. Download protein sequences from the *shroom* family of genes, and 
1. Prepare them for use in sequence alignment.  

For background on *shroom* and sequence alignments, see the previous post "Understanding the evolutionary history of a gene with phylogenetic trees."  In subsequent posts, I'll analyze these datas.

Most biologists download sequences directly from where they reside on internet, such as the **National Center For Biotechnology Information** (NCBI) websites.  In contrast, we'll use a function, `entrez_fretch()`, which accesses NCBI's [**Entrez**](ncbi.nlm.nih.gov/search/) system of database (ncbi.nlm.nih.gov/search/).  This function is from the `rentrez` package, which stands for "R-Entrez."

To work through the code in this post and use `entrez_fetch()`, you'll need to download the package `rentrez`

```{r, eval = F}
install.packages("rentrez")
```

We need to tell `entrez_fetch()` several things

1. `db = ...` the type of entrez database.  
1. `id = ...` the **accession** (ID) number of the sequence
1. `rettype = ...` file type what we want the function to return.  

Formally, these things are called **arguments** by *R*.

We'll use these settings:

1. `db = "protein"` to access the Entrez database of protein sequences  
1. `rettype = "fasta"`, which is a standard file format for nucleic acid and protein sequences

We'll set `id = ...` to sequences whose **accession numbers** are:

1. NP_065910: Human shroom 3
1. AAF13269:  Mouse shroom 3a
1. CAA58534:  Human shroom 2
1. XP_783573:  Sea urchin shroom

<!-- There are two highly conserved regions of shroom 3 -->

<!-- 1. ASD 1: aa 884 to aa 1062 in  hShroom3 -->
<!-- 1. ASD 2: aa 1671 to aa 1955 in hShroom3 -->

Normally we'd have to download these sequences by hand through pointing and clicking on GeneBank records on the NCBI website.  In *R* we can do it automatically; this might take a second.

All the code needed is this:

```{r}
# Human shroom 3 (H. sapiens)
hShroom3 <- rentrez::entrez_fetch(db = "protein", 
                          id = "NP_065910", 
                          rettype = "fasta")
```

<!-- Note: if you aren't connected to wifi or the server itself is having  problem, then when you run this code you may get an error like this: -->

<!-- `Quitting from lines 244-259 (23-MSA-walkthrough-shroom.Rmd)` -->
<!-- `Error: HTTP failure: 500` -->
<!-- `{"error":"error forwarding request","api-key":"71.182.228.80","type":"ip",` -->
<!-- `"status":"ok"}` -->
<!-- `Execution halted` -->

<!-- You may have to try again later to knit the code. -->

The output is in `FASTA` format.  We can look at the raw output by calling up the object we created.  This will run as single continuous string of characters without line breaks

```{r, eval = F}
hShroom3
```

```{r, echo = F}
paste(substr(hShroom3,1, 75),"...")
```

We can use the `cat()` function to do a little formatting for us; it essentially just enforces the **lines breaks**:

```{r}
cat(hShroom3)
```

Note the initial `>`, then the header line of `NP_065910.3 protein Shroom3 [Homo sapiens]`.  After that is the amino acid sequence.  The underlying data also includes the **newline character** `\n` to designate where each line of amino acids stops (that is, the location of line breaks).


We can get the rest of the data by just changing the `id = ...` argument in the calls to `entrez_fetch()`:

```{r}
# Mouse shroom 3a (M. musculus)
mShroom3a <- rentrez::entrez_fetch(db = "protein", 
                          id = "AAF13269", 
                          rettype = "fasta")

# Human shroom 2 (H. sapiens)
hShroom2 <- rentrez::entrez_fetch(db = "protein", 
                          id = "CAA58534", 
                          rettype = "fasta")


# Sea-urchin shroom
sShroom <- rentrez::entrez_fetch(db = "protein", 
                          id = "XP_783573", 
                          rettype = "fasta")
```

Here, I've pasted the function I used above three times into the code chunk and changed the id = ... statement.  Later in this script will avoid this clunky type of coding by using **for loops**.

I'm going to check about how long each of these sequences is - each should have an at least slightly different length. If any are identical, I might have repeated an accession name or re-used an object name. The function `nchar()` counts of the number of characters in an *R* object.

```{r}
nchar(hShroom3)
nchar(mShroom3a)
nchar(sShroom)
nchar(hShroom2)
```



## Prepping macromolecular sequences

>"90% of data analysis is data cleaning"  
~Just about every data analyst and data scientist

We have our sequences, but the current format isn't directly usable for us yet because there are several things that aren't sequence information

1. Metadata (the header)
1. Page formatting information (the newline character)

We can remove this non-sequence information using a function I wrote called `fasta_cleaner()`, which is in the [`compbio4all`](https://brouwern.github.io/compbio4all/index.html) package. The function uses **regular expressions** to remove the info we don't need.

If you had trouble downloading the `compbio4all` package function you can add `fasta_cleaner()` to your `R` session directly by running this code:

```{r}
fasta_cleaner <- function(fasta_object, parse = TRUE){

  fasta_object <- sub("^(>)(.*?)(\\n)(.*)(\\n\\n)","\\4",fasta_object)
  fasta_object <- gsub("\n", "", fasta_object)

  if(parse == TRUE){
    fasta_object <- stringr::str_split(fasta_object,
                                       pattern = "",
                                       simplify = FALSE)
  }

  return(fasta_object[[1]])
}

```


If we run the name of the command with out any quotation marks we can see the code:

```{r}
fasta_cleaner
```



Now use the function to clean our sequences; don't worry about what `parse = ...` is for right now.

```{r}
hShroom3  <- fasta_cleaner(hShroom3,  parse = F)
mShroom3a <- fasta_cleaner(mShroom3a, parse = F)
hShroom2  <- fasta_cleaner(hShroom2,  parse = F)
sShroom   <- fasta_cleaner(sShroom,   parse = F)
```

I wanted to do something four times, so I've repeated the same line of code four times with the necessary change.  This gets the job done, but there are better ways to do this using `for()` loops.

Now let's take a peek at what our sequences look like:

```{r, eval = F}
hShroom3
```

```{r, echo = F}
paste(substr(hShroom3,1,75),"...")
```



The header and slash-n newline characters are gone.  The sequence is now ready for use by our `R` alignment functions.

<!-- start second blog post -->

## Aligning pairs of sequences

We can do a [**global alignment**](https://tinyurl.com/y4du73zq) of one sequence against another using the `pairwiseAlignment()` function from the **Bioconductor** package `Biostrings` (note that capital "B" in `Biostrings`; most *R* package names are all lower case, but not this one).  Global alignment algorithms identify the best way to line up to sequences so that that optimal number of bases or amino acids are the same and the number of **indels** (insertions/deletions) are minimized.  (Global alignment contrasts with **local alignment**, which works with portions of sequences and is used in database search programs like **BLAST**, the Basic Local Alignment Search Tool used by many biologists).

Let's align human versus mouse shroom using the global alignment function `pairwiseAlignment():`

```{r}
align.h3.vs.m3a <- Biostrings::pairwiseAlignment(
                  hShroom3,
                  mShroom3a)
```

We can peek at the alignment

```{r}
align.h3.vs.m3a
```

The **score** tells us how closely they are aligned; higher scores mean the sequences are more similar.  In general, perfect matches increase scores the most, and indels decrease scores.  

Its hard to interpret scores on their own so we can get the **percent sequence identity (PID)** (aka percent identical, proportion identity, proportion identical) using the `pid()` function.

```{r}
Biostrings::pid(align.h3.vs.m3a)
```

So, *shroom3* from humans (hShroom3) and *shroom3* from mice (mShroom3) are ~71% similar (at least using this particular method of alignment, and there are many ways to do this!)

What about human shroom 3 and human shroom 2?  Shroom is a **gene family**, and there are different versions of the gene within a genome.

```{r}
align.h3.vs.h2 <- Biostrings::pairwiseAlignment(
                  hShroom3,
                  hShroom2)
```

If you take a look at the alignment you can see there are a lot of indels

```{r}
align.h3.vs.h2
```


Check out the score itself using `score()`, which accesses it directly without all the other information.

```{r}
score(align.h3.vs.h2)
```

Its negative because there are a LOT of indels.

Now the percent sequence alignment with `pid()`:

```{r}
Biostrings::pid(align.h3.vs.h2)
```

So Human shroom 3 and Mouse shroom 3 are 71% identical, but Human shroom 3 and human shroom 2 are only 34% similar?  How does it work out evolutionarily that a human and mouse gene are more similar than a human and a human gene?  What are the evolutionary relationships among these genes within the shroom gene family?

An important note: indels contribute negatively to an alignment score, but aren't used in the most common calculations for PID.



